pkg/model/pki.go

```diff
@@ -30,65 +32,80 @@ type PKIModelBuilder struct {
 
 var _ fi.ModelBuilder = &PKIModelBuilder{}
 
+// Build is responsible for generating the pki assets for the cluster
 func (b *PKIModelBuilder) Build(c *fi.ModelBuilderContext) error {
 	{
-		// Keypair used by the kubelet
 		t := &fitasks.Keypair{
 			Name:      fi.String("kubelet"),
 			Lifecycle: b.Lifecycle,
-
-			Subject: "o=" + user.NodesGroup + ",cn=kubelet",
-			Type:    "client",
+			Subject:   "o=" + user.NodesGroup + ",cn=kubelet",
+			Type:      "client",
 		}
 		c.AddTask(t)
 	}
 
 	{
-		// Secret used by the kubelet
-		// TODO: Can this be removed... at least from 1.6 on?
-		t := &fitasks.Secret{
-			Name:      fi.String("kubelet"),
+		t := &fitasks.Keypair{
+			Name:      fi.String("kube-scheduler"),
 			Lifecycle: b.Lifecycle,
+			Subject:   "cn=" + user.KubeScheduler,
+			Type:      "client",
 		}
 		c.AddTask(t)
 	}
 
 	{
-		// Keypair used by the kube-scheduler
 		t := &fitasks.Keypair{
-			Name:      fi.String("kube-scheduler"),
+			Name:      fi.String("kube-proxy"),
 			Lifecycle: b.Lifecycle,
-
-			Subject: "cn=" + user.KubeScheduler,
-			Type:    "client",
+			Subject:   "cn=" + user.KubeProxy,
+			Type:      "client",
 		}
 		c.AddTask(t)
 	}
 
 	{
-		// Secret used by the kube-scheduler
-		// TODO: Can this be removed... at least from 1.6 on?
-		t := &fitasks.Secret{
-			Name:      fi.String("system:scheduler"),
+		t := &fitasks.Keypair{
+			Name:      fi.String("kube-controller-manager"),
 			Lifecycle: b.Lifecycle,
+			Subject:   "cn=" + user.KubeControllerManager,
+			Type:      "client",
 		}
 		c.AddTask(t)
 	}
 
-	{
-		// Keypair used by the kube-proxy
-		t := &fitasks.Keypair{
-			Name:      fi.String("kube-proxy"),
-			Lifecycle: b.Lifecycle,
-
-			Subject: "cn=" + user.KubeProxy,
-			Type:    "client",
+	// @check if we need to generate certificates for etcd peers certificates from a different CA
+	// @question i think we should use another KeyStore for this, perhaps registering a EtcdKeyStore given
+	// that mutual tls used to verify between the peers we don't was
+	// For clients assuming we are using etcdv3 is can switch on user authentication and map the common names for auth
+	if b.Cluster.Spec.EnableEtcdTLS {
+		alternativeNames := []string{
+			fmt.Sprintf("*.internal.%s", b.ClusterName()),
+			fmt.Sprintf("*.internal.%s", b.Cluster.Spec.DNSZone),
+			"localhost", "127.0.0.1"}
+		{
+			// @question should wildcard here instead of generating per node. If we ever provide the
+			// ability to resize the master, this will become a blocker
+			c.AddTask(&fitasks.Keypair{
+				AlternateNames: alternativeNames,
+				Lifecycle:      b.Lifecycle,
+				Name:           fi.String("etcd"),
+				Subject:        "cn=etcd",
+				Type:           "server",
+			})
+		}
+		// @TODO when we move to etcdv3 we should generate a client cert for Calico as well
```

Comment:
We probably should just say that if you want to use TLS etcd, you need to use calico with the "kubernetes" backend, not the "etcd" backend.  We can enforce that in validation.